raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -441,22 +428,23 @@ private void becomeVotedFollower(int candidateId, int epoch) throws IOException
     private void onBecomeFollower() {
         resetConnections();
 
+        final long currentTime = time.milliseconds();
+
         // After becoming a follower, we need to complete pending fetches so that
         // they can be resent to the leader without waiting for their expiration
-        purgatory.completeAll(null);
-
-        // TODO: we may only return the future when the append is committed not appended locally,
-        //       in which case we would change this logic
-        uncommittedAppends.clear();
+        fetchPurgatory.completeAll(1, currentTime);
 
         for (UnwrittenAppend unwrittenAppend: unwrittenAppends) {
-            if (!unwrittenAppend.isCancelled()) {
-                unwrittenAppend.fail(new NotLeaderOrFollowerException(
-                    "Append refused since this node is no longer the leader"));
+            if (!unwrittenAppend.future.isDone()) {
+                unwrittenAppend.fail(new NotLeaderOrFollowerException("Append refused since this node is no longer " +
+                        "the leader"));
             }
         }
         unwrittenAppends.clear();
 
+        // Clearing the append purgatory should complete all future exceptionally since this node is no longer the leader
+        appendPurgatory.completeAll(Long.MAX_VALUE, currentTime);
```

Comment:
The path after a leader change seems a little brittle. We depend on the check for the leader status to detect this case. I wonder if it would make sense to add a `completeAllExceptionally` method to `FuturePurgatory` so that failures are explicit?

Reply:
Yeah I think I agree, but `completeAllExceptionally` is not very easy to implement though without much messing around the KafkaPurgatory, I will try again and let you know how far I get.